Poll the slo file from s3 periodically for slo monitoring

rpeela8899 opened the pull request:
- Created the capabilty to fetch slo file from the s3 bucket in stage initially
- Created the functionality to automatically periodic refresh the slo file in s3 for every hour  


Note:

- I need to update the mount path in the devcontainer.json file inorder to run the project in a container
- Added a mount for the AWS credentials to fetch the config from local env

TODO

- Unite tests
- Figure out the way to handle multi-regions. multi-env

benjdauno committed:
Draft of endpoint volume categorizer

benjdauno committed:
Working custom processor

benjdauno committed:
Include custom processor at the end of the pipeline

benjdauno committed:
Processor queries chronosphere

benjdauno committed:
Update query to chronosphere

benjdauno committed:
Basic custom-otel-collector restart

benjdauno committed:
Add valid basic otel config

benjdauno committed:
Add functioning Dec. 5 reboot files

benjdauno committed:
Clean up old files 1

benjdauno committed:
Clean up old files 2

benjdauno committed:
Add gitignore and enovy config to docker

benjdauno committed:
Add basic README to the project

benjdauno committed:
Use default verbosity for debug exporter

benjdaun committed:
Add Draft of dev container

benjdaun committed:
go mod tidy

benjdaun committed:
components.go reformat

benjdaun committed:
Ensure read/write permissions for linux

benjdaun committed:
Add devcontainer

benjdaun committed:
Include rpc2 query in excluded endpoints

benjdaun committed:
refresh sampling rates every so often

benjdaun committed:
Fix module name

benjdaun committed:
Dependency update

benjdaun committed:
Add telemetrygen to the Dockerfile

benjdaun committed:
No need for mutex

benjdaun committed:
File organization

benjdauno committed:
Fix path in launch.json

benjdauno committed:
Get ocb properly configured

benjdauno committed:
Docs improvement

benjdauno committed:
Add Dockerfile for the custom collector

benjdauno committed:
Add some tooling to the dev container

benjdauno committed:
Get docker compose working with volume mounts

benjdauno committed:
Refine build UX

benjdauno committed:
Start thinking about performance

benjdauno committed:
Add goimports to devcontainer to get proper language support

benjdauno committed:
Add actual processing logic to the log-processor

benjdauno committed:
Add kafka capabilities to the local setup

benjdauno committed:
Merge pull request #1 from benjdauno/feat-add-kafka

Add kafka capabilities to the local setup

benjdauno committed:
Put the .devcontainer of the host network

benjdauno committed:
Add additional processors already used by Affirm

benjdauno committed:
Merge pull request #2 from benjdauno/add-filter-receiver

Add additional processors already used by Affirm

benjdauno committed:
Send logs through Envoy

benjdauno committed:
Merge pull request #3 from benjdauno/envoy-setup

Send logs through Envoy

benjdauno committed:
fix conflict between docker mounts and being able to git commit

benjdauno committed:
Get Go language help back

benjdauno committed:
Set up better logging

benjdauno committed:
Add basic implementation of the exclusions list

benjdauno committed:
Add command to run binary from Makefile

benjdauno committed:
Merge pull request #4 from benjdauno/add-exclusions-list

benjdauno committed:
Get default configs into better shape

benjdauno committed:
Merge pull request #5 from benjdauno/fix-exclusions-config

Get default configs into better shape

benjdauno committed:
Make local webservers simulate our production endpoints

benjdauno committed:
Use siege in the Makefile test

benjdauno committed:
Merge pull request #6 from benjdauno:SRE-2150-set-up-nginx-to-mock-our-apps

Make local webservers simulate our production endpoints

benjdauno committed:
Include the cloud_region in the queries to prometheus

benjdauno committed:
Merge pull request #7 from benjdauno/SRE-2159-consider-that-different-sampling-rates-may-apply-for-different-countries

Include the cloud_region in the queries to prometheus

benjdauno committed:
Set valid dependency versions for build

arjunnayak committed:
OBSERVE-2774 Add custom slo metrics emitter log processor

- Emits 2 metrics: general request rate and latency breaches
- Uses hardcoded slo configs, to be replaced with proper YAML file parsing
- Uses OTELs internal metric and metricprovider sdk to emit metrics similar to other open source processors

arjunnayak committed:
separate otel configs and makefile runs

arjunnayak committed:
read and validate yaml config

arjunnayak committed:
read and use slo yaml config. define custom unmarshal to handle durations

arjunnayak committed:
update slo yaml with more checkout endpoints

arjunnayak committed:
update metric names, clean up logs, add endpoint type label

arjunnayak committed:
switch to debug log when endpoint not found to avoid polluting logs

arjunnayak committed:
add test for non existant slo

arjunnayak committed:
add tests and rename log processor file

arjunnayak committed:
move common setup to helper, cleanup

arjunnayak committed:
rm unused helper

arjunnayak committed:
add unit test and coverage commands, rename telemetry gen commands

arjunnayak committed:
fix test make command

arjunnayak committed:
rename Slo to SLO

arjunnayak committed:
update log level fall back. pr feedback, cleanup

benjdauno committed:
Merge pull request #8 from benjdauno/OBSERVE-2774

OBSERVE-2774 Add slo metrics processor

Looking good

arjunnayak committed:
Add health check extension to otel collector–used in k8s

arjunnayak committed:
Build amd64 binary for running in k8s

arjunnayak committed:
Update dockerfile–add test slo.yaml to container, document expose port for metrics

arjunnayak committed:
Fix processor to successfully handle live data in stage

- correct status code field to response_code
- fixes endpoint type handling, it should default to http
- debug log records

arjunnayak committed:
Bump version to 0.0.13

arjunnayak committed:
clean up slo.yaml

arjunnayak committed:
add note to dockerfile

arjunnayak committed:
Move istio log extraction to a separate method with a new struct. Separate telemetry setup. Move some method declarations around to follow a vertical hierarchy

arjunnayak committed:
Update endpoint type detector to return error if no content type is present instead of defaulting to http

arjunnayak committed:
Update endpoint detection when endpoint type is unknown

- the processor will try to get the endpoint name even with an unknown endpoint type
- this way it can still observe the overall request metric, but the latency breach metric will be skipped

rpeela8899 committed:
Poll the slo file from s3 periodically for slo monitoring

arjunnayak commented on slometricsemitter/log_processor.go:
```diff
@@ -248,6 +319,7 @@ func (c *EndpointSLOConfig) UnmarshalYAML(unmarshal func(interface{}) error) err
 }
 
 func (sloMetricsProc *sloMetricsProcessor) readSLOConfigFile(sloConfigFile string) (SLOConfig, error) {
+	sloMetricsProc.logger.Error("Inside function readSLOConfigFile")
```
needed?

arjunnayak commented on slometricsemitter/log_processor.go:
```diff
@@ -49,14 +91,43 @@ func (sloMetricsProc *sloMetricsProcessor) Start(ctx context.Context, host compo
 	}
 
+	// Start periodic refresh every hour
+	go func() {
```
let's make this a separate function to keep the `Start` method simple

arjunnayak commented on slometricsemitter/log_processor.go:
```diff
@@ -39,6 +46,41 @@ func (sloMetricsProc *sloMetricsProcessor) Capabilities() consumer.Capabilities
 
+// FetchSLOConfigFromS3 fetches the SLO config file from S3 and writes it to the specified path
+func (sloMetricsProc *sloMetricsProcessor) FetchSLOConfigFromS3(ctx context.Context, bucketName, key, localPath string) error {
+	cfg, err := config.LoadDefaultConfig(ctx, config.WithRegion("us-east-1")) // Replace with actual region after testing
```
nit: rename `config` to `awsConfig` for clarity

arjunnayak commented on slometricsemitter/log_processor.go:
```diff
@@ -49,14 +91,43 @@ func (sloMetricsProc *sloMetricsProcessor) Start(ctx context.Context, host compo
 	sloMetricsProc.logger.Info("Starting log processor with config", zap.String("sloConfigFile", sloMetricsProc.config.SLOConfigFile))
 
+	// Fetch SLO config initially
+	bucketName := "affirm-stage-logs"
```
this should be in config, and the bucket should be affirm-stage-ops

arjunnayak reviewed (changes requested):
Good work Rohit–left some comments. 

I don't think we can merge and deploy this until we're able to get iam permissions loaded on the pod in cosmops, otherwise it will break

arjunnayak commented:
a few more notes:
- go ahead and address the PR feedback but we need to create the PR using the base master branch on `benjdaun**o**/access_log_sampling`
- we can just start with stage us/prod us. The code should just expect the file to be in the bucket specified

arjunnayak committed:
Update endpoint type handling to default to http.

- After testing, it appears content-type isn't strongly set in our http responses.
- It's better for us to try to capture metrics for those endpoints than missing them

arjunnayak committed:
bump to 0.0.14

arjunnayak commented on slometricsemitter/log_processor.go:
```diff
@@ -49,14 +91,43 @@ func (sloMetricsProc *sloMetricsProcessor) Start(ctx context.Context, host compo
 
+	// Fetch SLO config initially
+	bucketName := "affirm-stage-logs"
+	key := "logs/slo/slo.yaml"
```
a first draft file has been uploaded to `s3://affirm-stage-ops/slos/http.yaml` so let's run this fetch process for http and rpc2

arjunnayak committed:
rm noisy log, bump to v0.0.15

arjunnayak committed:
Merge pull request #10 from benjdauno/OBSERVE-2845

OBSERVE-2845 Implement various fixes from live stage testing

- Add health check extension to custom base otel collector build (it's used in our k8s deployment)
- Build amd64 binary for running on k8s nodes (the vast majority of our apps/infra run on linux amd64)
- Fixes processor code to handle live data in stage: response_code instead of status_code, endpoint type handling
- Added debug logs for easier debugging of live incoming log records

rpeela8899 committed:
Update teh code to pull two s3 files

rpeela8899 committed:
Merge branch 'main' into OBSERVE-2895

rpeela8899 committed:
fix conf

rpeela8899 committed:
Update timepicker to 1hr

rpeela8899 committed:
Fix file being there on startup

rpeela8899 committed:
Fix the path issue

rpeela8899 committed:
Add tests

rpeela8899 committed:
fix tests again
